### PR TITLE
remove color from length count

### PR DESCRIPTION
**Pull request by d-m-u**
It'd be cool if the length calculation here didn't include color codes.

also it's something that if it were right I expect Keenan might back

### CODE REVIEW

File: lib/more_core_extensions/core_ext/array/tableize.rb [hunk 1]
```diff
@@ -118,7 +118,7 @@ def format_row(row, widths, justifications)
       end
 
       def format_field(field, width, justification)
-        field = field.to_s.gsub(/\n|\r/, '').slice(0, width)
+        field = field.to_s.gsub(/\n|\r|\e\[[^m]+m/, '').slice(0, width)
```
Comment (Fryguy) [inline lib/more_core_extensions/core_ext/array/tableize.rb:121]: I thin this should be `|\\e[` (note the double backslash)

**Comment by Fryguy** [comment]
Can you show a before/after?  I can't really understand what the issue was (i.e was the column too long or too short).

File: lib/more_core_extensions/core_ext/array/tableize.rb (hunk 1 above)
Comment (Fryguy) [reply lib/more_core_extensions/core_ext/array/tableize.rb:121]: Oh no wait, this is looking for a literal escape char...yeah this is fine.

File: spec/core_ext/array/tableize_spec.rb [hunk 2]
```diff
@@ -34,6 +34,17 @@
         expect(test.tableize).to eq(expected)
       end
 
+      it 'removes color' do
+        test = [["Col1", "Col2"], ["Val1", "Val2"], ["\033[31mReally Really Long Value3\e[0m", "Value4"]]
+        expected = <<-EOF
+ Col1                               | Col2
+------------------------------------|--------
+ Val1                               | Val2
+ Really Really Long Value3          | Value4
```
Comment (Fryguy) [inline spec/core_ext/array/tableize_spec.rb:43]: I'm surprised this has a lot of right-side padding...I'd expect your change to make it look like the test on line 26.

Comment (Fryguy) [reply spec/core_ext/array/tableize_spec.rb:43]: and since your test is passing, then is the PR not working as expected?

**Comment by d-m-u** [comment]
hey @Fryguy and @kbrock, thanks for the review on this, could I please ask you to take another look?

**Comment by bdunne** [comment]
Do we want to remove the color completely, or just for the field length?

File: lib/more_core_extensions/core_ext/array/tableize.rb [hunk 3]
```diff
@@ -118,7 +119,7 @@ def format_row(row, widths, justifications)
...
```
Comment (kbrock) [inline lib/more_core_extensions/core_ext/array/tableize.rb:122]: please remove this. I would like the escape codes to actually print out

File: spec/core_ext/array/tableize_spec.rb [hunk 4]
```diff
@@ -34,6 +34,17 @@
         expect(test.tableize).to eq(expected)
       end
 
+      it 'removes color' do
```
Comment (kbrock) [inline spec/core_ext/array/tableize_spec.rb:37]: This looks great

**Comment by d-m-u** [comment]
it _was_ only line 121, until [the comment yesterday](https://github.com/ManageIQ/more_core_extensions/pull/87#discussion_r450935603)

and when I asked about it again, here's what I got: 
```
Jul 07 17:31
that method totally looks right though
```

File: lib/more_core_extensions/core_ext/array/tableize.rb (hunk 3 above)
Comment (d-m-u) [reply lib/more_core_extensions/core_ext/array/tableize.rb:122]: okay, so this is the line you said before that was the right one that you wanted me to leave in

File: spec/core_ext/array/tableize_spec.rb [hunk 5]
```diff
@@ -34,6 +34,17 @@
         expect(test.tableize).to eq(expected)
       end
 
+      it 'removes color from length' do
+        test = [["Col1", "Col2"], ["Val1", "Val2"], ["\033[31mVal\e[0m", "Value4"]]
+        expected = <<-EOF
+ Col1 | Col2
+------|--------
+ Val1 | Val2
+ Val  | Value4
```
Comment (NickLaMuro) [inline spec/core_ext/array/tableize_spec.rb:43]: So this is probably going to fail currently since if you are keeping the color code in the result, it will be "width'd" to the length of the raw string, but still display the color code in the output here.

Comment (NickLaMuro) [reply spec/core_ext/array/tableize_spec.rb:43]: Also, seems like your regexp still needs some work, since it seems to be deleting `"Val"` from the output:

https://travis-ci.com/github/ManageIQ/more_core_extensions/jobs/359385835#L278

Comment (kbrock) [reply spec/core_ext/array/tableize_spec.rb:43]:
```diff
-       expected: " Col1 | Col2\n------|--------\n Val1 | Val2\n Val  | Value4\n"
+            got: " Col1 | Col2\n------|--------\n Val1 | Val2\n \e[31 | Value4\n"
```

I was actually expecting:

```
            got: " Col1 | Col2\n------|--------\n Val1 | Val2\n \e[31mVal\e[0m | Value4\n"
```

File: lib/more_core_extensions/core_ext/array/tableize.rb [hunk 6]
```diff
@@ -118,7 +119,7 @@ def format_row(row, widths, justifications)
       end
 
       def format_field(field, width, justification)
-        field = field.to_s.gsub(/\n|\r/, '').slice(0, width)
+        field = field.to_s.gsub(/\n|\r/, '').slice(0, width) unless field.to_s.include?("\e")
```
Comment (Fryguy) [inline lib/more_core_extensions/core_ext/array/tableize.rb:122]: So, if the field had a \n, we won't remove it if it also has color?  I don't understand that.

Comment (Fryguy) [reply lib/more_core_extensions/core_ext/array/tableize.rb:122]: This will also break the max_width value when it's set really small... To verify, duplicate the test for really long values with max width set, and then add in some color.

In fact, when you do that I expect you are going to get broken field codes, but that can be remedied by adding in a reset `\e[0m;` when the field is truncated.

Comment (kbrock) [reply lib/more_core_extensions/core_ext/array/tableize.rb:122]: It looks like we use `max_width` in only one spot in `manageiq-release`. With a value of 75.

Someone having ANSI color escape sequences AND a small `max_width` is starting to go towards the probably never gonna happen side of edge cases.

Pulling the to_s and gsub onto a separate line seems simple enough.

Implementing max width and escape sequences requires some pretty heavy lifting. It is a very deep rabbit hole and can we let this bug come out once

**Comment by kbrock** [comment]
thanks. this looks nice @d-m-u

File: lib/more_core_extensions/core_ext/array/tableize.rb (hunk 6 above)
Comment (Fryguy) [reply lib/more_core_extensions/core_ext/array/tableize.rb]: The gem is generic and supports it as a parameter, so we need to account for it (I use this also for personal projects for example).

Comment (kbrock) [reply lib/more_core_extensions/core_ext/array/tableize.rb]: @Fryguy I'm not sure how to approach this in a way that will be understandable and will not be a bunch of code.

```ruby
rspec spec/core_ext/array/tabelize_spec.rb:37

# given widths = [4,6]
# and we're in the first column

format_field(field: "\033[31mVal\e[0m", width:4)
# => " \033[31mVal\e[0m "
```

|character|pos||
|---|---|---|
|V|5|
|a|6|
|l|7|
|.|12|no 4th character, end of string|

So the method would do a `field.slice(12)` since the 4th character is past the end of the string
Possibly we also want to tack on an extra `"\e[0m"` to be safe.

The only idea I have for this logic is to walk the string skipping escape sequences as we calculate the desired width.

Do you have any better ideas?
Do you think the benefit of this edge case outweighs the code we'll add?

**Comment by Fryguy** [comment]
@d-m-u Accidental close?

File: lib/more_core_extensions/core_ext/array/tableize.rb [hunk 7]
```diff
@@ -118,7 +119,8 @@ def format_row(row, widths, justifications)
       end
 
       def format_field(field, width, justification)
-        field = field.to_s.gsub(/\n|\r/, '').slice(0, width)
+        field = field.to_s.gsub(/\n|\r/, '')
+        field = field.slice(0, width) unless field.include?("\e")
```
Comment (Fryguy) [inline lib/more_core_extensions/core_ext/array/tableize.rb:123]: We have to clip the width to _something_ because otherwise any long field will go over the max_width. I think the following will work:

```ruby
field = field.to_s.gsub(/\n|\r/, '').slice(0, width)
filed << "\e[0m" if field.include?("\e")
```

This can be tested by duplicating the "with really long column value" and the "'with really long column value and :max_width option'' and having a color variant of those.  You may also want test variants where the max_width comes after an escape sequence, before an escape sequence, and directly inside an escape sequence.

**Comment by d-m-u** [comment]
@Fryguy no, we're going to implement it in benchmark-sweet and not need this dependency.

**Comment by Fryguy** [comment]
ok...I still like the feature, so I'll probably reopen at some point.

File: lib/more_core_extensions/core_ext/array/tableize.rb (hunk 7 above)
Comment (Fryguy) [reply lib/more_core_extensions/core_ext/array/tableize.rb:123]: So, yeah, I wrote this off the cuff, and it has issues, but the following seems to work...it's a bit brute forcey, but seems ok to me:

```ruby
      def format_field(field, width, justification)
        field = field.to_s.gsub(/\n|\r/, '')

        if field.include?("\e")
          while strip_ansi(field).size > width
            field.chop!
          end
          field << ANSI_RESET if field.include?("\e") && !field.end_with?(ANSI_RESET)
        else
          field = field.slice(0, width)
        end

        "%0#{justification}#{width}s" % field
      end

      def strip_ansi(field)
        field.gsub(ANSI_ESCAPE_SEQUENCE, '')
      end
```

**Comment by Fryguy** [comment]
Updated with the code snippet from https://github.com/ManageIQ/more_core_extensions/pull/87#discussion_r455286172 and added a bunch more tests for the edge cases.

@bdunne Please review.

**Comment by bdunne** [comment]
@Fryguy Can you add a test with multiple colors?

**Comment by Fryguy** [comment]
Updated with a better algorithm that determines matching escape sequences that fall into the expected width.  Still need to add the specs.

File: lib/more_core_extensions/core_ext/array/tableize.rb [hunk 8]
```diff
@@ -118,10 +121,42 @@ def format_row(row, widths, justifications)
...
 
       def format_field(field, width, justification)
-        field = field.to_s.gsub(/\n|\r/, '').slice(0, width)
+        field = field.to_s.gsub(/\n|\r/, '')
+        field = ansi_truncate(field, width)
         "%0#{justification}#{width}s" % field
       end
 
+      def ansi_escapes?(field)
+        field.match?(ANSI_ESCAPE_SEQUENCE)
```
Comment (bdunne) [inline lib/more_core_extensions/core_ext/array/tableize.rb:130]: What do you think of `!!(field =~ ANSI_ESCAPE_SEQUENCE)` to get the tests passing on older rubies?
